ably/realtime/connection.py

```diff
@@ -0,0 +1,84 @@
+import logging
+import asyncio
+import websockets
+import json
+from ably.util.exceptions import AblyAuthException
+from enum import Enum
+
+log = logging.getLogger(__name__)
+
+
+class ConnectionState(Enum):
+    INITIALIZED = 'initialized'
+    CONNECTING = 'connecting'
+    CONNECTED = 'connected'
+    CLOSING = 'closing'
+    CLOSED = 'closed'
+    FAILED = 'failed'
+
+
+class RealtimeConnection:
+    def __init__(self, realtime):
+        self.options = realtime.options
+        self.__ably = realtime
+        self.__state = ConnectionState.INITIALIZED
+        self.__connected_future = None
+        self.__websocket = None
+
+    async def connect(self):
+        if self.__state == ConnectionState.CONNECTED:
+            return
+
+        if self.__state == ConnectionState.CONNECTING:
+            if self.__connected_future is None:
+                log.fatal('Connection state is CONNECTING but connected_future does not exits')
+                return
+            await self.__connected_future
+        else:
+            self.__state = ConnectionState.CONNECTING
+            self.__connected_future = asyncio.Future()
+            asyncio.create_task(self.connect_impl())
+            await self.__connected_future
+            self.__state = ConnectionState.CONNECTED
+
+    async def close(self):
+        self.__state = ConnectionState.CLOSING
+        if self.__websocket:
+            await self.__websocket.close()
+        else:
+            log.warn('Connection.closed called while connection already closed')
+        self.__state = ConnectionState.CLOSED
+
+    async def connect_impl(self):
+        async with websockets.connect(f'wss://{self.options.realtime_host}?key={self.ably.key}') as websocket:
+            self.__websocket = websocket
+            task = asyncio.create_task(self.ws_read_loop())
+            await task
+
+    async def ws_read_loop(self):
+        while True:
+            raw = await self.__websocket.recv()
+            msg = json.loads(raw)
+            action = msg['action']
+            if action == 4:  # CONNECTED
```

Comment:
I recognise we may consider this [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) at this stage, as an atomic increment to this codebase, however might it not be worthwhile defining these in an `Enum` from the outset?

Reply:
I see this was addressed in https://github.com/ably/ably-python/pull/311/commits/940d0cc49c6754f3c48d6479e9a4e38eac1ebfb8 - thanks.